crates/nu-command/src/core_commands/do_.rs

```diff
@@ -117,6 +118,26 @@ impl Command for Do {
                 span,
                 metadata,
             }) if capture_errors => {
+                // These streams must always be captured and in this specific order to
+                // prevent deadlocks.
+                let mut stderr_ctrlc = None;
+                let stderr_msg = match stderr {
+                    None => "".to_string(),
+                    Some(stderr_stream) => {
+                        stderr_ctrlc = stderr_stream.ctrlc.clone();
+                        stderr_stream.into_string().map(|s| s.item)?
+                    }
+                };
+
+                let mut stdout_ctrlc = None;
+                let stdout_msg = match stdout {
```

Comment:
you can delegate the stdout stream consuming login in another thread, then we can avoid deadlock

Reply:
You can check my pr(#7234) and do something similar:
```rust
                let stdout_handler = stdout.map(|stdout_stream| {
                    std::thread::spawn(move || {
                        let stdout_ctrlc = stdout_stream.ctrlc.clone();
                        let stdout_span = stdout_stream.span;
                        let stdout_bytes = match stdout_stream.into_bytes() {
                            Err(_) => vec![],
                            Ok(bytes) => bytes.item,
                        };
                        RawStream::new(
                            Box::new(vec![Ok(stdout_bytes)].into_iter()),
                            stdout_ctrlc,
                            stdout_span,
                        )
                    })
                });

               // then consuming stderr message

                ....
                ....

                let stdout = if let Some(h) = stdout_handler {
                    match h.join() {
                        Err(err) => {
                            return Err(ShellError::ExternalCommand(
                                "Fail to receive external commands stdout message".to_string(),
                                format!("{err:?}"),
                                span,
                            ))
                        }
                        Ok(res) => Some(res),
                    }
                } else {
                    None
                };
```

Reply:
Thanks! I still think the stream framework needs to be rewritten so this bug is impossible, but this is a good stopgap solution.